samples/web/content/apprtc/js/call.js

```diff
@@ -145,56 +142,83 @@ Call.prototype.toggleAudioMute = function() {
-  // TODO(jiayl): We don't need to wait for the signaling channel to open before
-  // start signaling.
-  var channelPromise = this.channel_.open().catch(function(error) {
-    this.onError_('WebSocket open error: ' + error.message);
+  this.registerWithRoomServer_(roomId).then(function(roomParams) {
+    // TODO(tkchin): clean up response format. JSHint doesn't like it.
+    /* jshint ignore:start */
+    this.params_ = {};
+    this.params_.clientId = roomParams.client_id;
+    this.params_.isInitiator = roomParams.is_initiator === 'true';
+    this.params_.isLoopback = roomParams.is_loopback === 'true';
+    this.params_.roomId = roomParams.room_id;
+    this.params_.roomLink = roomParams.room_link;
+    this.params_.mediaConstraints = parseJSON(roomParams.media_constraints);
+    this.params_.offerConstraints = parseJSON(roomParams.offer_constraints);
+    this.params_.peerConnectionConfig = parseJSON(roomParams.pc_config);
+    this.params_.peerConnectionConstraints = parseJSON(roomParams.pc_constraints);
+    this.params_.turnRequestUrl = roomParams.turn_url;
+    this.params_.turnTransports = roomParams.turn_transports;
+    this.params_.audioSendBitrate = roomParams.asbr;
+    this.params_.audioSendCodec = roomParams.audio_send_codec;
+    this.params_.audioRecvBitrate = roomParams.arbr;
+    this.params_.audioRecvCodec = roomParams.audio_receive_codec;
+    this.params_.opusMaxPbr = roomParams.opusmaxpbr;
+    this.params_.opusFec = roomParams.opusfec;
+    this.params_.videoSendBitrate = roomParams.vsbr;
+    this.params_.videoSendInitialBitrate = roomParams.vsibr;
+    this.params_.videoSendCodec = roomParams.video_send_codec;
+    this.params_.videoRecvBitrate = roomParams.vrbr;
+    this.params_.videoRecvCodec = roomParams.video_receive_codec;
+    this.params_.wssUrl = roomParams.wss_url;
+    this.params_.wssPostUrl = roomParams.wss_post_url;
+     /* jshint ignore:end */
+    this.params_.messages = roomParams.messages;
+    
+    if (this.params_.isLoopback) {
+      setupLoopback();
+    }
+  }.bind(this)).catch(function(error) {
+    this.onError_('Room server register error: ' + error.message);
     return Promise.reject(error);
-  }.bind(this));
-
-  // Asynchronously register with GAE.
-  var registerPromise =
-      this.registerWithRoomServer_().then(function(roomParams) {
-        // The only difference in parameters should be clientId and isInitiator,
-        // and the turn servers that we requested.
-        // TODO(tkchin): clean up response format. JSHint doesn't like it.
-        /* jshint ignore:start */
-        this.params_.clientId = roomParams.client_id;
-        this.params_.roomId = roomParams.room_id;
-        this.params_.isInitiator = roomParams.is_initiator === 'true';
-        /* jshint ignore:end */
-        this.params_.messages = roomParams.messages;
+  }.bind(this)).then(function() {
+    var mediaPromise = this.maybeGetMedia_();
+    var turnPromise = this.maybeGetTurnServers_();
```

Comment:
Requesting media and TURN are slow, so we want to start them as soon as possible instead of waiting for 'register'. 
We could/should create the media constraints locally from the URL params, instead of getting them from the server.
But for TURN, we have to get the default turn url from the initial page load params.